### PR TITLE
Initial vue support

### LINKED ISSUE #5: Feature request: Vue components
Using [vue-component-compiler](https://github.com/vuejs/vue-component-compiler).

### PR DESCRIPTION
**Pull request by DeMoorJasper**
Basic vue support using `vue-compiler-utils`

- [x] Have a working demo/test in place
- [x] Compile vue files using `vue-compiler-utils`
- [x] Support pre-processors
- [x] Sourcemaps support
- [x] Add more tests

Closes #5

### CODE REVIEW

**Comment by laosb** [comment]
It seems parcel try to install some packages, which makes parcel itself not useable (missing prelude.min.js) on the first run. I need to stop parcel & patch parcel after the first run.

And, vue styles are not correctly packaged into css files

**Comment by laosb** [comment]
Styles in `.vue`s wasn't generated to css bundles at all (I'm using dynamic imports to import vue components), and I don't really know why those `import imgUrl from './xxx.jpg'` doesn't work in this branch.

**Comment by laosb** [comment]
The intended behavior would be:

Bundler extract CSS (scoped if needed, with or without preprocessors) and JavaScript (Template & script) from `.vue` files, and bundle them into a `.css` file and a `.js` file. When a vue component is imported, it should import both `.css` & `.js` file.

**Comment by devongovett** [comment]
Ok here's what I think we should do. I think we don't really want to use the default vue compiler since it inlines everything into the JS (e.g. styles), and has its own compilation pipeline for styles, scripts, and templates. We should use the pipeline already built in parcel to do this as it will allow for correct file watching, css extraction, bundle splitting, etc.

What we want to do is parse the `.vue` files into their component pieces (script, style, template), and pass the resulting pieces through parcel's existing infrastructure. We should get most of what we want for free this way.

In order to do this, we'll need a couple of changes. We need a way to compose asset types together. Currently this is kinda faked in some asset types like stylus by extending from another asset type, and calling `super` to get that asset type's behavior in addition to some overrides. Ideally, parcel would just handle this automatically. If the stylus asset just compiled and output CSS from its `generate` method, then parcel could just pass that to the CSS asset type to do further compilation automatically - no class extension necessary. Similarly, the Vue asset type could just parse the file into its component pieces (script, style, template) and do no other work - parcel would take care of passing each of those to the correct compilers (other asset types).

**Comment by DeMoorJasper** [comment]
Alright i'll change this to generate script, style and template seperately and once Devon's composition system lands I'll update it to use it

**Comment by devongovett** [comment]
Implemented pipelines in #1065. Changed the base of this PR to point to that one, and merged in the changes. Updated VueAsset to use the new architecture. It could probably be a little cleaner still, but that should happen in the next major version when we can break the API.

There are still a few things to fix, but things seem to be mostly working now. Compilation happens, including support for preprocessors like Stylus, CoffeeScript, Jade (not implemented yet but theoretically), etc. I was able to successfully run the `vue-dependencies`, and `vue-preprocessors` apps from the tests.

screen shot 2018-03-25 at 12 04 21 am

Still need to fix:

- [ ] source maps
- [ ] HMR
- [ ] Scoped styles
- [ ] Custom blocks
- [ ] Functional templates? (idk what that is)

@DeMoorJasper feel free to work on any of those if you like. 😄

File: src/assets/VueAsset.js [hunk 1]
```diff
@@ -8,11 +8,14 @@ class VueAsset extends Asset {
   }
 
   async generate() {
+    // Is being used in component-compiler-utils, errors if not installed...
+    await localRequire('vue-template-compiler', this.name);
```
[the remark this answers is not in the document]
Comment (DeMoorJasper) [reply src/assets/VueAsset.js]: It probably should, see https://github.com/vuejs/component-compiler-utils/pull/3

[the remark this answers is not in the document]
Comment (devongovett) [reply src/assets/VueAsset.js]: awesome, thanks

**Comment by laosb** [comment]
Tested and found something strange: All scoped styles are not correctly processed except one. I'll need to investigate further.

**Comment by devongovett** [comment]
@laosb is your app open source?

**Comment by devongovett** [comment]
Ok implemented all of the missing features except custom blocks. I'm not sure how to support those as they might require some configuration, so going to leave them out of the initial version.

Tried it out on the Vue Hacker News example app, and it seems to work well. HMR and code splitting work out of the box. You can see a diff from the original version here: https://github.com/devongovett/vue-hackernews-2.0/pull/1. Lots of webpack boilerplate removed, along with tons of dependencies. Pretty cool! 😉 

screen shot 2018-03-25 at 8 58 37 pm

**Comment by gnijuohz** [comment]
This is really awesome! 👍

**Comment by laosb** [comment]
@devongovett Unfortunately it's private so I can't just post a repo link there. By the way, do you mean those components with `<slot>` in it by saying "custom blocks"? I discovered that all components with `<slot>` in are not properly styled.

**Comment by laosb** [comment]
Tested a lot and see dynamic importing is not correctly working to me. When I try to use dynamic Vue component by `component: { SomeComponent: () => import('./xxx') }`, dynamic css bundles (generated from `<style scoped>` in `xxx.vue`), which should be also loaded, were not loaded (but it did generate).

And, all bundles are actually loaded on first page load. Is this intended?

I will try to put together a minimal reproduction ASAP. Anyway, good job 👍

**Comment by laosb** [comment]
Something is quite annoying is, Parcel doesn't always output the same when given the same input. I removed `vue-hot-reload-api` from my package.json, then all of sudden all scoped styles failed. After my revert, it doesn't work anymore, even if I tried to clear `.cache` and `dist`. What a strange behavior it is.

**Comment by marcelkorpel** [comment]
I don't want to clutter this PR with building questions, but I'm not sure whether or not I did the right thing to build Parcel using this PR locally. First, I cloned the repository, then I fetched this fork using

    git clone https://github.com/parcel-bundler/parcel.git
    cd parcel
    git fetch origin pull/1052/head:origin/vue-v3

Then I installed all dependencies and built Parcel using Yarn:

    yarn install

After creating an alias I tried to build [shawwn's boiler plate](https://github.com/parcel-bundler/parcel/issues/5#issuecomment-359168173), but I (still) only get an empty `#app` div. What am I doing wrong?

BTW, `yarn test` failed 1 test:

> Error: Failed to install postcss-cssnext.

**Comment by marcelkorpel** [comment]
Wait, I still had to checkout the branch, in which these commits already are pushed. However, after trying to clone and build Parcel again, I still get the same issue.

**Comment by DeMoorJasper** [comment]
@laosb would be nice if you could provide a test repo for the bugs you encounter so we can test them or add them to the vue test cases.
@marcelkorpel This branch has nothing to do with shawwwn's example or implementation, his example should however work as all the test cases are build around his hello world test atm.

**Comment by marcelkorpel** [comment]
@DeMoorJasper I understand about the example, but somewhat more complicated code I wrote exhibited the same error. I also tried to use `npm run dev`, as that tends to give different results, but whatever I do, only an (untouched) `#app` div is present after opening localhost:1234.

**Comment by gnijuohz** [comment]
@marcelkorpel I tried shawwn's parcel-simple and it worked for me. Although I had to `yarn install vue` first.

**Comment by marcelkorpel** [comment]
Thanks for the tip, @gnijuohz, but I already tried that.

But now I see this error that I got earlier, when using Parcel from npm and before I installed parcel-plugin-vue: in Firefox, the console complains

> InvalidCharacterError: String contains an invalid character

In Chromium, the console is a bit more descriptive:

>Uncaught DOMException: Failed to execute 'createElement' on 'Document': The tag name provided ('/a6698109b5cf167c75bf419931e81341.vue') is not a valid name.

The code pointed to is
```js
function isUnknownElement (tag) {
…
tag = tag.toLowerCase();
/* istanbul ignore if */
if (unknownElementCache[tag] != null) {
  return unknownElementCache[tag]
}
var el = document.createElement(tag);
```
But I thought you don't need parcel-plugin-vue anymore when using the vue-v3 branch, am I right?

**Comment by laosb** [comment]
@devongovett Ah I see. Hopefully it's not widely used, at least I haven't use it before. We can safely consider it later.

@DeMoorJasper I pushed [my test app](https://github.com/laosb/parcel-vue-test) to GitHub, currently it works, if you need, you may add it to tests. It includes several ways to import a component in Vue, and all of them has failed at least once during my test.

@marcelkorpel Try to clean `.cache`, `node_modules` and `dist`, then retry install Parcel and build it. It should works. I've seen this before and it's how I solved it.

**Comment by laosb** [comment]
So now it sounds like we are going to see Vue support released?

**Comment by DeMoorJasper** [comment]
@laosb I'm pretty sure it'll be shipped in the next release (1.7)

**Comment by devongovett** [comment]
🎉

**Pull request merged**